File: src/test/java/edu/kit/iti/formal/stvs/model/expressions/TestExpressionParser.java
package edu.kit.iti.formal.stvs.model.expressions;

import edu.kit.iti.formal.stvs.model.expressions.parser.UnsupportedExpressionException;
import edu.kit.iti.formal.stvs.model.expressions.parser.ExpressionParser;
import edu.kit.iti.formal.stvs.model.expressions.parser.ParseException;
import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;

/**
 * Created by philipp on 17.01.17.
 */
public class TestExpressionParser {

  // Initialize parser with context information:
  private final String columnName = "X";
  private final ExpressionParser parser = new ExpressionParser(columnName);

  public TestExpressionParser() {
  }

  private Expression assertParseExpressionEqual(String expr, Expression expected)
      throws ParseException, UnsupportedExpressionException {
    Expression parsedExpr = parser.parseExpression(expr);
    System.out.println(expr + " <means> " + parsedExpr);
    assertEquals(
        expected,
        parsedExpr
    );
    return parsedExpr;
  }

  @Test
  public void testConstants() throws ParseException, UnsupportedExpressionException {
    // TRUE means X = TRUE
    assertParseExpressionEqual("TRUE", SimpleExpressions.equal(SimpleExpressions.var(columnName), SimpleExpressions.literal(true)));
    // both upper and lower case
    assertParseExpressionEqual("false", SimpleExpressions.equal(SimpleExpressions.var(columnName), SimpleExpressions.literal(false)));
    // 1 means X = 1
    assertParseExpressionEqual("1", SimpleExpressions.equal(SimpleExpressions.var(columnName), SimpleExpressions.literal(1)));
    // - means TRUE
    assertParseExpressionEqual("-", SimpleExpressions.literal(true));
  }

  @Test
  public void testVariables() throws ParseException, UnsupportedExpressionException {
    assertParseExpressionEqual("b", SimpleExpressions.equal(SimpleExpressions.var(columnName), SimpleExpressions.var("b")));
    assertParseExpressionEqual("b = ! FALSE", SimpleExpressions.equal(SimpleExpressions.var("b"), SimpleExpressions.not(SimpleExpressions.literal(false))));
  }

  @Test
  public void testVariablesWithIndex() throws ParseException, UnsupportedExpressionException {
    assertParseExpressionEqual("A[-2]", SimpleExpressions.equal(SimpleExpressions.var(columnName), SimpleExpressions.var("A", -2)));
    assertParseExpressionEqual("(A[0])", SimpleExpressions.var("A", 0));
  }

  @Test(expected = UnsupportedExpressionException.class)
  public void testVariablesOnlyNegativeIndex1() throws ParseException, UnsupportedExpressionException {
    parser.parseExpression("A[2]");
  }

  @Test(expected = UnsupportedExpressionException.class)
  public void testVariablesOnlyNegativeIndex2() throws ParseException, UnsupportedExpressionException {
    parser.parseExpression("(A[2])");
  }

  @Test
  public void testOnesided() throws ParseException, UnsupportedExpressionException {
    // < 2 means X < 2
    assertParseExpressionEqual("< 2", SimpleExpressions.lessThan(SimpleExpressions.var(columnName), SimpleExpressions.literal(2)));
  }

  @Test
  public void testBinaryOperators() throws ParseException, UnsupportedExpressionException {
    Map<String, BinaryFunctionExpr.Op> binaryOps = new HashMap<>();
    binaryOps.put("+", BinaryFunctionExpr.Op.PLUS);
    binaryOps.put(" - ", BinaryFunctionExpr.Op.MINUS);
    binaryOps.put("*", BinaryFunctionExpr.Op.MULTIPLICATION);
    binaryOps.put("/", BinaryFunctionExpr.Op.DIVISION);
    // power is commented out in grammar for some reason ?
    //binaryOps.put("**", BinaryFunctionExpr.Op.POWER);
    binaryOps.put("%", BinaryFunctionExpr.Op.MODULO);
    binaryOps.put(" MOD ", BinaryFunctionExpr.Op.MODULO);

    binaryOps.put(">", BinaryFunctionExpr.Op.GREATER_THAN);
    binaryOps.put("<", BinaryFunctionExpr.Op.LESS_THAN);
    binaryOps.put(">=", BinaryFunctionExpr.Op.GREATER_EQUALS);
    binaryOps.put("<=", BinaryFunctionExpr.Op.LESS_EQUALS);

    binaryOps.put("=", BinaryFunctionExpr.Op.EQUALS);
    binaryOps.put("!=", BinaryFunctionExpr.Op.NOT_EQUALS);
    binaryOps.put("<>", BinaryFunctionExpr.Op.NOT_EQUALS);

    binaryOps.put("&", BinaryFunctionExpr.Op.AND);
    binaryOps.put(" AND ", BinaryFunctionExpr.Op.AND);
    binaryOps.put("|", BinaryFunctionExpr.Op.OR);
    binaryOps.put(" OR ", BinaryFunctionExpr.Op.OR);
    binaryOps.put(" XOR ", BinaryFunctionExpr.Op.XOR);
    binaryOps.put(" xor ", BinaryFunctionExpr.Op.XOR);

    for (Map.Entry<String, BinaryFunctionExpr.Op> binaryOperationEntry : binaryOps.entrySet()) {
      String operator = binaryOperationEntry.getKey();
      BinaryFunctionExpr.Op operation = binaryOperationEntry.getValue();

      assertParseExpressionEqual("2" + operator + "2",
          new BinaryFunctionExpr(operation, SimpleExpressions.literal(2), SimpleExpressions.literal(2))
      );
    }
  }

  @Test
  public void testUnaryOperators() throws ParseException, UnsupportedExpressionException {
    // parens enforce a "expression" rule (not single-sided or constant or blah)
    assertParseExpressionEqual("(- (2))", SimpleExpressions.negate(SimpleExpressions.literal(2)));
    assertParseExpressionEqual("(!TRUE)", SimpleExpressions.not(SimpleExpressions.literal(true)));
    assertParseExpressionEqual("(NOT TRUE)", SimpleExpressions.not(SimpleExpressions.literal(true)));
  }

  @Test(expected = ParseException.class)
  public void testInvalidPlus() throws ParseException, UnsupportedExpressionException {
    parser.parseExpression("= +3");
  }

  @Test
  public void testInterval() throws ParseException, UnsupportedExpressionException {
    // [1, 4] means X >= 1 AND X <= 4
    assertParseExpressionEqual("[4, 1]",
        SimpleExpressions.and(SimpleExpressions.greaterEqual(SimpleExpressions.var(columnName), SimpleExpressions.literal(4)), SimpleExpressions.lessEqual(SimpleExpressions.var(columnName), SimpleExpressions.literal(1)))
    );
  }

  @Test(expected = ParseException.class)
  public void testInvalidInterval() throws ParseException, UnsupportedExpressionException {
    parser.parseExpression("[1,3");
  }

  @Test(expected = UnsupportedExpressionException.class)
  public void testUnsupportedFunctioncall() throws ParseException, UnsupportedExpressionException {
    parser.parseExpression("function(1, 2)");
  }

  @Test(expected = UnsupportedExpressionException.class)
  public void testUnsupportedGuardedIf() throws ParseException, UnsupportedExpressionException {
    parser.parseExpression("if :: TRUE -> FALSE :: FALSE -> TRUE fi");
  }

  @Test
  public void testRecognizeEnum() throws ParseException, UnsupportedExpressionException {
    TypeEnum colorsEnum = new TypeEnum("COLORS", Arrays.asList("red", "green", "blue"));
    Set<Type> typeContext = new HashSet<>();
    typeContext.add(colorsEnum);

    parser.setTypeContext(typeContext);
    assertParseExpressionEqual("(blue)", SimpleExpressions.literalEnum("blue", colorsEnum));
    //assertParseExpressionEqual("red", equal(var(cellName), literalEnum("red", colorsEnum)));
  }

}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/model/expressions/TestExpressionParser.kt
package edu.kit.iti.formal.stvs.model.expressions

import edu.kit.iti.formal.stvs.model.expressions.SimpleExpressions.equal
import edu.kit.iti.formal.stvs.model.expressions.SimpleExpressions.lessThan
import edu.kit.iti.formal.stvs.model.expressions.SimpleExpressions.literal
import edu.kit.iti.formal.stvs.model.expressions.SimpleExpressions.variable
import edu.kit.iti.formal.stvs.model.expressions.parser.ExpressionParser
import edu.kit.iti.formal.stvs.model.expressions.parser.ParseException
import edu.kit.iti.formal.stvs.model.expressions.parser.UnsupportedExpressionException
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import kotlin.test.assertFailsWith

/**
 * Created by philipp on 17.01.17.
 */
class TestExpressionParser {
    // Initialize parser with context information:
    private val columnName = "X"
    private val parser = ExpressionParser(columnName)

    @Throws(ParseException::class, UnsupportedExpressionException::class)
    private fun assertParseExpressionEqual(expr: String, expected: Expression): Expression {
        val parsedExpr = parser.parseExpression(expr)
        println("$expr <means> $parsedExpr")
        assertEquals(expected, parsedExpr)
        return parsedExpr
    }

    @Test
    fun testConstants() {
        // TRUE means X = TRUE
        assertParseExpressionEqual("TRUE", equal(variable(columnName), literal(true)))
        // both upper and lower case
        assertParseExpressionEqual("false", equal(variable(columnName), literal(false)))
        // 1 means X = 1
        assertParseExpressionEqual("1", equal(variable(columnName), literal(1)))
        // - means TRUE
        assertParseExpressionEqual("-", literal(true))
    }

    @Test
    fun testVariables() {
        assertParseExpressionEqual("b", equal(variable(columnName), variable("b")))
        assertParseExpressionEqual(
            "b = ! FALSE",
            equal(variable("b"), SimpleExpressions.not(literal(false)))
        )
    }

    @Test
    @Throws(ParseException::class, UnsupportedExpressionException::class)
    fun testVariablesWithIndex() {
        assertParseExpressionEqual("A[-2]", equal(variable(columnName), variable("A", -2)))
        assertParseExpressionEqual("(A[0])", variable("A", 0))
    }

    @Test
    fun testVariablesOnlyNegativeIndex1() {
        assertFailsWith<UnsupportedExpressionException> { parser.parseExpression("A[2]") }
    }

    @Test
    fun testVariablesOnlyNegativeIndex2() {
        assertFailsWith<UnsupportedExpressionException> { parser.parseExpression("(A[2])") }
    }

    @Test
    fun testOnesided() {
        // < 2 means X < 2
        assertParseExpressionEqual("< 2", lessThan(variable(columnName), literal(2)))
    }

    @Test
    @Throws(ParseException::class, UnsupportedExpressionException::class)
    fun testBinaryOperators() {
        val binaryOps = hashMapOf(
            "+" to BinaryFunctionExpr.Op.PLUS,
            " - " to BinaryFunctionExpr.Op.MINUS,
            "*" to BinaryFunctionExpr.Op.MULTIPLICATION,
            "/" to BinaryFunctionExpr.Op.DIVISION,
            // power is commented out in grammar for some reason ?
            //binaryOps.put("**", BinaryFunctionExpr.Op.POWER);
            "%" to BinaryFunctionExpr.Op.MODULO,
            " MOD " to BinaryFunctionExpr.Op.MODULO,
            ">" to BinaryFunctionExpr.Op.GREATER_THAN,
            "<" to BinaryFunctionExpr.Op.LESS_THAN,
            ">=" to BinaryFunctionExpr.Op.GREATER_EQUALS,
            "<=" to BinaryFunctionExpr.Op.LESS_EQUALS,
            "=" to BinaryFunctionExpr.Op.EQUALS,
            "!=" to BinaryFunctionExpr.Op.NOT_EQUALS,
            "<>" to BinaryFunctionExpr.Op.NOT_EQUALS,
            "&" to BinaryFunctionExpr.Op.AND,
            " AND " to BinaryFunctionExpr.Op.AND,
            "|" to BinaryFunctionExpr.Op.OR,
            " OR " to BinaryFunctionExpr.Op.OR,
            " XOR " to BinaryFunctionExpr.Op.XOR,
            " xor " to BinaryFunctionExpr.Op.XOR
        )

        for ((operator, operation) in binaryOps) {
            assertParseExpressionEqual(
                "2" + operator + "2",
                BinaryFunctionExpr(operation, literal(2), literal(2))
            )
        }
    }

    @Test
    @Throws(ParseException::class, UnsupportedExpressionException::class)
    fun testUnaryOperators() {
        // parens enforce an "expression" rule (not single-sided or constant or blah)
        assertParseExpressionEqual("(- (2))", SimpleExpressions.negate(literal(2)))
        assertParseExpressionEqual("(!TRUE)", SimpleExpressions.not(literal(true)))
        assertParseExpressionEqual("(NOT TRUE)", SimpleExpressions.not(literal(true)))
    }

    @Test
    fun testInvalidPlus() {
        assertFailsWith<ParseException> { parser.parseExpression("= +3") }
    }

    @Test
    fun testInterval() {
        // [1, 4] means X >= 1 AND X <= 4
        assertParseExpressionEqual(
            "[4, 1]",
            SimpleExpressions.and(
                SimpleExpressions.lessEqual(
                    literal(4), variable(columnName)
                ), SimpleExpressions.lessEqual(variable(columnName), literal(1))
            )
        )
    }

    @Test()
    fun testInvalidInterval() {
        assertFailsWith<ParseException> { parser.parseExpression("[1,3") }
    }

    @Test
    fun testUnsupportedFunctioncall() {
        assertFailsWith<UnsupportedExpressionException> {
            parser.parseExpression("function(1, 2)")
        }
    }

    @Test
    fun testUnsupportedGuardedIf() {
        parser.parseExpression("if :: TRUE -> FALSE :: FALSE -> TRUE fi")
    }

    @Test
    fun testRecognizeEnum() {
        val colorsEnum = TypeEnum("COLORS", mutableListOf("red", "green", "blue"))
        val typeContext: MutableSet<Type> = HashSet()
        typeContext.add(colorsEnum)

        parser.setTypeContext(typeContext)
        assertParseExpressionEqual("(blue)", SimpleExpressions.literalEnum("blue", colorsEnum))
        //assertParseExpressionEqual("red", equal(var(cellName), literalEnum("red", colorsEnum)));
    }

    @Test
    fun testLongIntegerLiteral0() {
        assertFailsWith<NumberFormatException> { parser.parseExpression("1324574839294857") }
    }

    @Test
    fun testLongIntegerLiteral1() {
        parser.parseExpression("" + (Short.MAX_VALUE + 1))
    }

    @Test
    fun testLongIntegerLiteral2() {
        parser.parseExpression("" + (Short.MIN_VALUE - 1))
    }
}
